api/envoy/extensions/outlier_detection_monitors/consecutive_errors/v3/consecutive_errors.proto

```diff
@@ -22,13 +22,11 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // If number of consecutive errors exceeds the threshold, monitor will report that the host
 // is unhealthy.
 message ConsecutiveErrors {
+  // Base monitor config. Defines which results should be interpreted as errors
+  // and whether to enforce ejection when number of errors exceeds defined threshold.
+  common.v3.MonitorCapture errors = 1 [(validate.rules).message = {required: true}];
```

Comment:
Because you have made a big change to the proto. So, I want to do a quick discussion about this:

Should this field be part of the protocol-specific router if necessary? I mean the result of request is put by the router filter (db router or http router, etc), outlier detection basically only need to know if the request is error or not and neen't care the specific implementation or status codes/db results.

By this way, the router take the responsibility to map the specific status to the `Result` (`ExtOriginRequestFailed`, `ExtOriginRequestSuccess`). The outlier extension take the responsibility to count and check.

Reply:
> Should this field be part of the protocol-specific router (if necessary)? I mean the result of request is put by the router filter (db router or http router, etc) into the outlier detenction module. 

I do not think it should be part of protocol-specific router API. Each protocol-specific router will use its corresponding type. So, for example http router will use `HttpCode` to report the result of HTTP transaction, but http router will never use `LocalOriginResult` nor `DatabaseResult`. Similarly, redis terminal router, will use `DatabaseResult` to report how the transaction with Redis database completed, but will never use `HttpCode`. So, in other words structures used by each protocol specific router as hardcoded in each router and do not have to be part of its API.

> The outlier detection basically only need to know if the request is error or not and neen't care the specific implementation or status codes/db results.

That is correct. And user has ability to define which responses should be treated as errors. For example, for http, in old implementation only 500 codes were treated as errors and now user may define any range to be treated as errors: 4xx, 5xx.

> By this way, the router take the responsibility to map the specific status to the Result (ExtOriginRequestFailed, ExtOriginRequestSuccess). The outlier extension take the responsibility to count and check.

The second part is correct. The extension counts errors and determines if it should mark host as unhealthy or not. I implemented only consecutive errors extension for now, but there will be others. The first part, is somewhat correct, but I would like to depart from using "old" enums like `ExtOriginRequestFailed`, `ExtOriginRequestSuccess` and solely use new structures: `HttpCode`, `LocalOriginEvent` and `DatabaseResult`. For now there is some mapping for backwards compatibility.

> By this way, even we want to add new status/protocol support, we needn't change the outlier detection's API or implementation? This make the outlier detection could be protocol independent in some way? (Of course, the HTTP always be special could get spectial treatmet.)

No changes will be required to extensions. You would only need to add a new proto messages (or reuse existing message (`HttpCode`, `LocalOriginEvent` or `DatabaseResult`)) and define within that message what is considered an error and what is considered a success. Extensions are independent of types. They only care if whatever is reported is error or success, and this is implemented within `HttpCode`, `LocalOriginEvent` and `DatabaseResult`, so you are right when you say that outlier detection is protocol independent.

Reply:
The `errors` field is used to determine whether a status is error or not. But I think this acutally isn't outlier detection's work. This should be done by the protocol-specific router (if necessary). So, I say the `errors` may be part of Router rather than placing it in the outlier detection.

> No changes will be required to extensions. You would only need to add a new proto messages (or reuse existing message (HttpCode, LocalOriginEvent or DatabaseResult)) and define within that message what is considered an error and what is considered a success.

I think new protocol support (for example, dubbo) need to change the `MonitorCapture` and the the signature of the `putResult` method of outlier detection to provide their `DubboCode` (or similar things). It acutally means lots of code change. (We needn't to change both the core `DetectorHostMonitor` and all extensions.)

So, I think the expected way to make the outlier detection protocol independent is let the protocol-specific router to handle all the protocol-specific things. And the outlier detection only handle the protocol-independent result. (The `Result` now, HTTP is special again, so additional code is used.)

Reply:
> The common.v3.MonitorCapture errors field is used to determine whether a status is error or not. But I think this acutally isn't outlier detection's work. This should be done by the protocol-specific router (if necessary). So, I say the errors may be part of protocol specific router (of course, router only require their protocol's config, for example, the HTTP router only need HttpCodes) rather than placing a compositing error buckets in the outlier detection.

Theoretically, such separation makes a lot of sense, but based on my experience with outlier, it would also limit possible use cases. There are few reasons, why I think moving `MonitorCapture` to procol-specific filter should be avoided:
- backwards compatibility would be harder to achieve. Currently, there are no changes to filters. I just package http-code or local origin event into Result and forward to extensions.
- You can create several monitors where the same code may mean different thing. So, the http-code or event should be categorized as error or not-error only when it reaches the extensions. Protocol specific filter should only report what happened.
- Even if `MonitorCapture` was moved to protocol-specific filter's API, events generated when something goes wrong may not come only from that filter. For example http router runs on top of TCP code. So, you could define in the router's API that say 5xx are errors, but if upstream closes the connection, `LocalOriginEvent` will be sent, not http code. And that `LocalOriginEvent` is not generated by the router.

> In current implementation, I think new protocol support (for example, dubbo) need to change the MonitorCapture and the the signature of the putResult method of outlier detection to provide their DubboCode (or similar things). It acutally means lots of code change. (We needn't to change both the core DetectorHostMonitor and all extensions.) And any new protocol support will add similar changes again.

Yes, some changes are required, but minimal. I prototyped solution for https://github.com/envoyproxy/envoy/issues/24215, adding literally 5-10 lines. Redis filter recognized the error, packaged it into `DatabaseResult` and extension matched it against bucket of 3 consecutive Redis errors and marked the redis host as unhealthy. It was very, very minimal. Notice that extension's code `onMatch`, `onReset`, `onSuccess` are completely protocol independent.

Reply:
I need to say thanks all your time first and sorry I raised this discussion so late. But I believe most work of this PR is great and only this interface need more discussion.

Reply:
> Yes, some changes are required, but minimal. I prototyped solution for https://github.com/envoyproxy/envoy/issues/24215, adding literally 5-10 lines. Redis filter recognized the error, packaged it into DatabaseResult and extension matched it against bucket of 3 consecutive Redis errors and marked the redis host as unhealthy. It was very, very minimal. Notice that extension's code onMatch, onReset, onSuccess are completely protocol independent.

I am afraid it's not minimal. Given the redis or dubbo as an example (DB support is not implemented now, so, anyway). Then these changes are necessary:

1. New API to represent the new type status code.
2. Update the putResult() method signature to support new type status code for both core Monitor and all extensions. (This change effect our core code which is what I want to avoid when we doing new protocol outlier detection support.)
3. Handle the status in the extensions.

And this bring another risk to all custom extensions. Give an exist extension, it may handle the status by following way:

```
// Assume the status is a variant.
if(absl::hold_althernative<HttpCode>(status) {
  // ...
} else {
  const auto default_type = absl::get<Result>(status);
  // ...
}
```

If you add new status support, the code will be broken.

Reply:
> backwards compatibility would be harder to achieve. Currently, there are no changes to filters. I just package http-code or local origin event into Result and forward to extensions.

I didn't see any backwards compatibility problem? All current filters map their status to the Result with a hardcoded behavior (except HTTP). Only the one who want more flexible status mapping need to add new API to control it's mapping.

> You can create several monitors where the same code may mean different thing. So, the http-code or event should be categorized as error or not-error only when it reaches the extensions. 

This should be avoid. Same code shouldn't have different meaning in different extensions. I don't think this is one of our target scenarios.

Reply:
> Protocol specific filter should only report what happened.

I think this is most important core of our this discussion.

See, if the filter only report what happended, then, the outlier detection could never be protocol independent. Because it need to detect different protocols' status code.

If outlier detection couldn't be protocol independent, then that means we cannot provide the outlier detection support to the non-core protocols because the outlier detection self is the core feature of envoy and shouldn't depends on non-core features.

For example, there are lots of contrib extensions for various protocols, if they wants outlier detection support, how can we do it? It's impossible to add their status to the outlier detection's API. We still need a common abstraction and let these contrib extensions to map their status to the common abstraction. I think this is also why we designed the `Result`. (May be it's not pefect, but is more closer to our expected way.)